### PR TITLE
#2339 - Cannot assign string to property of type MongoDB\BSON\ObjectIdInterface

### PR DESCRIPTION
**Pull request by smilesrg**
https://github.com/doctrine/mongodb-odm/issues/2339

|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | no
| Issue |  2339

#### Summary

Added test for a problem "Cannot assign string to property of type MongoDB\BSON\ObjectIdInterface"

### CODE REVIEW

**Comment by smilesrg** [comment]
@malarzm would you please approve this PR for a CI testing?

**Comment by smilesrg** [comment]
@malarzm once again, please? :-)

**Comment by malarzm** [comment]
It's running :)

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH2339Test.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+use function phpversion;
+use function version_compare;
+
+class GH2339Test extends BaseTest
+{
+    public function setUp(): void
+    {
+        if (version_compare((string) phpversion(), '8.0', '<')) {
+            self::markTestSkipped('PHP 8.0 is required to run this test');
```
Comment (malarzm) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH2339Test.php]: Why is PHP 8.0 required for this test?

Comment (smilesrg) [reply tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH2339Test.php]: Because of property typehinting `ObjectIdInterface`

Comment (smilesrg) [reply tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH2339Test.php]: Whoops, I was confused. Updating requirements to PHP 7.4

**Comment by smilesrg** [comment]
So, failing test is ready:

```
1) Doctrine\ODM\MongoDB\Tests\Functional\Ticket\GH2339Test::testObjectIdInterfaceInEmbeddedDocuments
TypeError: Cannot assign string to property Doctrine\ODM\MongoDB\Tests\Functional\Ticket\GH2339Test\ParentDocument::$id of type MongoDB\BSON\ObjectIdInterface

/home/runner/work/mongodb-odm/mongodb-odm/vendor/doctrine/persistence/lib/Doctrine/Persistence/Reflection/TypedNoDefaultReflectionProperty.php:56
/home/runner/work/mongodb-odm/mongodb-odm/lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadata.php:1453
/home/runner/work/mongodb-odm/mongodb-odm/lib/Doctrine/ODM/MongoDB/UnitOfWork.php:1076
/home/runner/work/mongodb-odm/mongodb-odm/lib/Doctrine/ODM/MongoDB/UnitOfWork.php:1674
/home/runner/work/mongodb-odm/mongodb-odm/lib/Doctrine/ODM/MongoDB/UnitOfWork.php:1631
/home/runner/work/mongodb-odm/mongodb-odm/lib/Doctrine/ODM/MongoDB/DocumentManager.php:458
/home/runner/work/mongodb-odm/mongodb-odm/tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH2339Test.php:32

```

**Comment by smilesrg** [comment]
@malarzm Workflows are still waiting for approval... would you please approve? :-)

**Comment by smilesrg** [comment]
@malarzm thanks! Test have failed, as expected https://github.com/doctrine/mongodb-odm/pull/2342/checks?check_run_id=3189958004

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH2339Test.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+use Doctrine\ODM\MongoDB\Tests\Functional\Ticket\GH2339Test\ParentDocument;
+
+use function phpversion;
+use function version_compare;
+
+class GH2339Test extends BaseTest
+{
+    public function setUp(): void
+    {
+        if (version_compare((string) phpversion(), '7.4', '<')) {
```
Comment (IonBazan) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH2339Test.php:18]: Please use `@requires` annotation instead (see #2360). You can use that on the class-level in this case.

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH2339Test.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            self::markTestSkipped('PHP 7.4 or higher is required to run this test');
+        }
+
+        parent::setUp();
+    }
+
+    public function testObjectIdInterfaceInEmbeddedDocuments()
```
Comment (IonBazan) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH2339Test.php:25]:
```suggestion
    public function testObjectIdInterfaceInEmbeddedDocuments(): void
```

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH2339Test/ParentDocument.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+     * @var InlinedDocument[]
+     */
+    protected array $inlined = [];
+
+    public function getId(): ObjectIdInterface
+    {
+        return $this->id;
+    }
+
+    public function addInlined(InlinedDocument $document)
```
Comment (IonBazan) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket/GH2339Test/ParentDocument.php:30]: We require at least PHP 7.2 so we can safely add `void` type hint.
```suggestion
    public function addInlined(InlinedDocument $document): void
```

**Comment by malarzm** [comment]
@IonBazan these changes can be applied while working on a fix for the failing test case. Sadly I had no time yet to do so :(

**Comment by alcaeus** [comment]
Closing this: as per https://github.com/doctrine/mongodb-odm/issues/2339#issuecomment-914063872, this is intended behaviour (even though it's confused me in the beginning as well).

**Pull request closed without merging**